devutils/update_lists.py

```diff
@@ -205,84 +205,106 @@ def _check_regex_match(file_path, search_regex):
+    used_pep_list is a list of PRUNING_EXCLUDE_PATTERNS that have been matched
+    used_pip_list is a list of PRUNING_INCLUDE_PATTERNS that have been matched
+    used_dep_list is a list of DOMAIN_EXCLUDE_PREFIXES that have been matched
+    used_dip_list is a list of DOMAIN_INCLUDE_PATTERNS that have been matched
+    domain_sub_list is a list of files that require domain substitution
+    pruning_list is a list of file to be pruned
+    symlink_list is a list of symlinks
+    """
+    if not path.is_file():
+        # NOTE: Path.rglob() does not traverse symlink dirs; no need for special handling
+        return
+    relative_path = path.relative_to(_source_tree)
+    if path.is_symlink():
+        try:
+            resolved_relative_posix = path.resolve().relative_to(_source_tree).as_posix()
+        except ValueError:
+            # Symlink leads out of the source tree
+            return
+        symlink_list.append((resolved_relative_posix, relative_path.as_posix()))
+        return
+    try:
+        if should_prune(path, relative_path, used_pep_list, used_pip_list):
+            pruning_list.append(relative_path.as_posix())
+        elif should_domain_substitute(path, relative_path, used_dep_list, used_dip_list):
+            domain_sub_list.append(relative_path.as_posix())
+    except: #pylint: disable=bare-except
+        get_logger().exception('Unhandled exception while processing %s', relative_path)
+
+
+def compute_lists():
     """
     Compute the binary pruning and domain substitution lists of the source tree.
-    Returns a tuple of two items in the following order:
+    Returns a tuple of three items in the following order:
     1. The sorted binary pruning list
     2. The sorted domain substitution list
-
-    source_tree is a pathlib.Path to the source tree
-    search_regex is a compiled regex object to search for domain names
+    3. An UnusedPatterns object
     """
-    pruning_set = set()
-    domain_substitution_set = set()
-    deferred_symlinks = dict() # POSIX resolved path -> set of POSIX symlink paths
-    source_tree = source_tree.resolve()
+    # Set up the multiprocess lists
+    manager = Manager()
+    used_pep_list = manager.list() # pruning_exclude_patterns
+    used_pip_list = manager.list() # pruning_include_patterns
+    used_dep_list = manager.list() # domain_exclude_prefixes
+    used_dip_list = manager.list() # domain_include_patterns
+    domain_sub_list = manager.list()
+    pruning_list = manager.list()
+    symlink_list = manager.list() # POSIX resolved path -> set of POSIX symlink paths
+
+    # Launch multiple processes iterating over the source tree
+    with Pool() as procpool:
+        procpool.starmap(
+            compute_lists_proc,
+            zip(
+                _source_tree.rglob('*'), repeat(used_pep_list), repeat(used_pip_list),
+                repeat(used_dep_list), repeat(used_dip_list), repeat(domain_sub_list),
+                repeat(pruning_list), repeat(symlink_list)))
```

Comment:
We both want to improve performance here, so I am wondering if `multiprocessing`'s Managers is the right synchronization mechanism to use. I'm looking a bit into Managers, but I'm still not clear on the details. For example, what underlying IPC mechanism do Managers use to share data?

But first, let's be clear on the higher-level design. Looking at the code, I believe these two invariants are true:

1. Each worker process uses data from the main process without modifying them.
2. Each worker process only needs to write to its output lists; it never needs to read its output.

Based on these invariants, we can conclude that **each worker process can work independently from each other**. 

Now let's consider your use of Manager. Whenever a worker modifies a list, the Manager needs to employ a locking mechanism to prevent a race condition with other worker processes, right? However, this is very inefficient because we just proved that worker processes don't need to share data among themselves (something that a Manager is useful for). In other words, workers should be free to run completely independently from each other (i.e. without synchronization) until they are ready to return their output.

Therefore, we should consider another synchronization mechanism. For example, can we simply return the lists at the end of each worker function and aggregate them in the main process?